File: Java/LeetCode/src/main/java/com/example/leetcode/rotateRight.java
package com.example.leetcode;

import java.util.List;

/**
 * @author huangchunchen
 * @date 2021/9/15 11:39
 * @description
 * 61. Rotate List
 * Given the head of a linked list, rotate the list to the right by k places.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: head = [1,2,3,4,5], k = 2
 * Output: [4,5,1,2,3]
 * Example 2:
 *
 *
 * Input: head = [0,1,2], k = 4
 * Output: [2,0,1]
 *
 *
 * Constraints:
 *
 * The number of nodes in the list is in the range [0, 500].
 * -100 <= Node.val <= 100
 * 0 <= k <= 2 * 109
 */
public class rotateRight {
    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(5);
        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        System.out.println(rotateRight1(node1, 2));
    }

    public static ListNode rotateRight(ListNode head, int k) {
        if (k == 0 || head == null || head.next == null) {
            return head;
        }
        int n = 1;
        ListNode iter = head;
        while (iter.next != null) {
            iter = iter.next;
            n++;
        }
        int add = n - k % n;
        if (add == n) {
            return head;
        }
        iter.next = head;
        while (add-- > 0) {
            iter = iter.next;
        }
        ListNode ret = iter.next;
        iter.next = null;
        return ret;
    }

    public static ListNode rotateRight1(ListNode head, int k) {
        int count = 0;
        ListNode temp = head;
        while (temp!=null){
            temp = temp.next;
            count++;
        }
        k = k%count;
        ListNode fast = head;
        ListNode slow = head;

        while (k-->0){
            fast = fast.next;
        }

        while (fast.next!=null){
            fast = fast.next;
            slow = slow.next;
        }
        fast.next = head;
        head = slow.next;
        slow.next=null;
        return head;
    }
}

File: cdd/LeetCode/LeetCode/rotateRight.cpp
#include <stdio.h>
#include "ListNode.cpp"

using namespace std;

class Solution {
public:
	ListNode* rotateRight(ListNode* head, int k) {
		if (k==0||head ==NULL || head->next == NULL)
		{
			return head;
		}

		int count = 1;
		ListNode* temp = head;
		while (temp->next != NULL)
		{
			temp = temp->next;
			count++;
		}

		int len = count - k % count;
		if (len == count)
		{
			return head;
		}

		temp->next = head;

		while (len-- > 0) {
			temp = temp->next;
		}

		ListNode* res = temp->next;
		temp->next = NULL;
		return res;
	}

	ListNode* rotateRight1(ListNode* head, int k) {
		if (k==0||head==NULL||head->next==NULL)
		{
			return head;
		}
		int count = 0;
		ListNode* temp = head;
		while (temp != NULL)
		{
			temp = temp->next;
			count++;
		}

		k = k % count;
		ListNode* fast = head;
		ListNode* slow = head;

		while (k-- > 0) {
			fast = fast->next;
		}

		while (fast->next != NULL)
		{
			fast = fast->next;
			slow = slow->next;
		}

		fast->next = head;
		head = slow->next;
		slow->next = NULL;
		return head;
	}
};